### PR TITLE
Throw a clear exception if OutputModule SelectEvents refers to a non-existent Path or Process

### LINKED ISSUE #886: Add test and improve error message when OutputModule SelectEvents refers to a Process that does not exist in history
See https://github.com/cms-sw/cmssw/issues/44743

### LINKED ISSUE #44743: Add test and improve error message when OutputModule SelectEvents refers to a Process that does not exist in history
The `StdException` failure reported in https://cms-talk.web.cern.ch/t/issues-with-run-379404/39046/1 was caused by an OutputModule `SelectEvents` to refer to a Process for which a `TriggerResults` data product did not exist. The exception message was just
```
----- Begin Fatal Exception 14-Apr-2024 18:15:47 CEST-----------------------
An exception of category 'StdException' occurred while
   [0] Calling EventProcessor::runToCompletion (which does almost everything after beginJob and before endJob)
Exception Message:
A std::exception was thrown.
vector::_M_range_check: __n (which is 4294967295) >= this->size() (which is 14)
----- End Fatal Exception -------------------------------------------------
```

We should make the exception message more clear, and also add tests for this case.

### LINKED ISSUE #44744: Job gets stuck if `OutputModuleCore::prePrefetchSelection()` throws an exception
In a similar situation as in https://github.com/cms-sw/cmssw/issues/44743, the following job gets stuck (in starvation of TBB tasks to run)
```py
import FWCore.ParameterSet.Config as cms

process = cms.Process("TEST")

process.source = cms.Source("EmptySource")
process.maxEvents.input = 3

process.out = cms.OutputModule("SewerModule",
    SelectEvents = cms.untracked.PSet(
        SelectEvents = cms.vstring("nonexistent_path:NONEXISTENT_PROCESS")
    ),
    name = cms.string("out"),
    shouldPass = cms.int32(0)
)

process.intprod = cms.EDProducer("IntProducer", ivalue=cms.int32(3))

process.ep = cms.EndPath(process.intprod+process.out)

# process.add_(cms.Service('ZombieKillerService', secondsBetweenChecks=cms.untracked.uint32(10)))
```

In the `ep` the `intprod` must be before `out` for the job to get stuck. Other way around the job terminates properly with an exception.

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This was already the case for a non-existent Path of the present process. For non-existent Path of an earlier Process, or a non-existent Process, lead to a confusing exception message, and in some cases the job got stuck. An example of such stuck job was having an unrelated module in front of the OutputModule in the same EndPath.

Resolves https://github.com/cms-sw/cmssw/issues/44743
Resolves https://github.com/cms-sw/cmssw/issues/44744
Resolves https://github.com/cms-sw/framework-team/issues/886

#### PR validation:

Added unit tests pass

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@Dr15Jones Please review

**Comment by antoniovilela** [comment]
Is full_cmssw advisable here?

**Comment by makortel** [comment]
> Is full_cmssw advisable here?

We can certainly do that, thanks for the suggestion. I was in any case waiting for https://github.com/cms-sw/cmssw/pull/44787 to be merged before launching next round of tests (not much point testing before)

**Comment by makortel** [comment]
test parameters:
   - full_cmssw = true

**Comment by makortel** [comment]
> There are some workflows for which there are errors in the baseline:
> 29661.402 step 3
> 29661.403 step 3
> 29834.402 step 2
> 29834.403 step 2

@smuzaffar These extra workflows were not run as part of the tests of this PR (or as part of IBs AFAICT). Are the failures reported in the test summary above perhaps because some other PR used the same baseline, and those extra workflows were requested for the tests of that PR?

**Comment by makortel** [comment]
Seems like this PR would be good to go. @Dr15Jones could you take a last look?

**Comment by smuzaffar** [comment]
> Are the failures reported in the test summary above perhaps because some other PR used the same baseline, and those extra workflows were requested for the tests of that PR?

yes looks like these extra workflows were requested by https://github.com/cms-sw/cmssw/pull/44874#issuecomment-2084923103. Bot should have ignore these workflow failures for this PR tests. I will look in to this

File: FWCore/Framework/interface/WorkerInPath.h [hunk 1]
```diff
@@ -109,7 +109,7 @@ namespace edm {
                                     typename T::TransitionInfoType const& info,
                                     ServiceToken const& token,
                                     StreamID streamID,
-                                    typename T::Context const* context) {
+                                    typename T::Context const* context) noexcept {
```
Comment (Dr15Jones) [inline FWCore/Framework/interface/WorkerInPath.h:112]: Should we also mark `Worker::doWorkAsync` and `Worker::doWorkNoPrefetchingAsync` as `noexcept` as this routine calls them?

Comment (makortel) [reply FWCore/Framework/interface/WorkerInPath.h:112]: As we discussed, I opened an issue about that (seemed too deep of a rabbit hole to be done as part of this PR).

**Comment by makortel** [comment]
+core

**Comment by rappoccio** [comment]
+1

**Comment by smuzaffar** [comment]
> yes looks like these extra workflows were requested by [#44874 (comment)](https://github.com/cms-sw/cmssw/pull/44874#issuecomment-2084923103). Bot should have ignore these workflow failures for this PR tests. I will look in to this

https://github.com/cms-sw/cms-bot/pull/2221 should fix this for future PR tests

**Pull request merged**